Web Symbols: improve and extend documentation. (#1039)
JetBrains/intellij-sdk-docs: topics/reference_guide/custom_language_support/websymbols_integration.md

<!-- Copyright 2000-2023 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license. -->

# Web Symbols Integration with language features

<link-summary>
How to integrate your Web Symbols with a particular language feature.
</link-summary>

IDEs provide built-in support for Web Symbols integration with main language features of HTML, CSS and JavaScript.
You can contribute static symbols through Web Types or Custom Elements Manifest, or you can
create and register [`WebSymbolsQueryConfigurator`](%gh-ic%/platform/webSymbols/src/com/intellij/webSymbols/query/WebSymbolsQueryConfigurator.kt)
extension.

> Your implementation of `WebSymbolsQueryConfigurator` needs to override `getScope` method, in which you
> create a list of `WebSymbolScope`s depending on the provided `PsiElement` (`element` parameter) and `WebSymbolContext` (`context` parameter).
> The below list of supported language features specify information on what kind of `PsiElement`s to expect for each of
> the supported language features.

Reviewer: ```suggestion
The implementation of `WebSymbolsQueryConfigurator` requires overriding the `getScope` method. This process involves the creation of a list of `WebSymbolScope`s, depending on the provided `PsiElement` (`element` parameter) and `WebSymbolContext` (`context` parameter). The following list of supported language features outlines the types of `PsiElement`s that can be expected for each supported language feature.```

## Supported language features

### HTML

#### Elements {#html-elements}

> |---|---|

Reviewer 2: we shouldn't use a table for two pieces of information

suggestion:

```
**Namespace**: `html`

**Kind**: `elements`
```

or similar, apply to all on this page

Author: Hmm, the table looks much nicer graphically, I've experimented with different approaches, including your suggestion, but table looked the best for me. Have you experimented and you're sure your suggestion looks the best for you?

|||
|Namespace|`html`|
|Kind|`elements`|

Web Symbols, which represent available HTML elements.

> HTML element symbols can be contributed statically or dynamically (through `WebSymbolsQueryConfigurator`).

Reviewer 2: `[`WebSymbolsQueryConfigurator`](%gh-ic%/platform/webSymbols/src/com/intellij/webSymbols/query/WebSymbolsQueryConfigurator.kt)`

The statically contributed symbols are available globally, depending on the context setting of the contributing Web Types file.

> For dynamic contributions the `WebSymbolsQueryConfigurator#getScope`'s `element` parameter can be:

Reviewer 2: use `.` instead of `#` for delimiting method names

> For dynamic contributions the `WebSymbolsQueryConfigurator#getScope`'s `element` parameter can be:

Reviewer: ```suggestion
For dynamic contributions the `WebSymbolsQueryConfigurator.getScope()`'s `element` parameter can be:
```

> - `HtmlTag` - Web Symbols should represent available HTML elements. *The HTML tag actual name should not be taken into account when building the scope.*

Reviewer 2: `[`HtmlTag`](%gh-ic%/xml/xml-psi-api/src/com/intellij/psi/html/HtmlTag.java)`

- `CssElement` - Web Symbols should represent available HTML elements within a particular CSS file.

The matched Web Symbol is taken as a scope for matching HTML attributes of the tag.

#### Attributes {#html-attributes}

|---|---|
|||
|Namespace|`html`|
|Kind|`attributes`|

> Web Symbols, which represent available HTML attributes.

Reviewer 2: Web Symbols representing...

HTML attribute symbols can be contributed statically or dynamically (through `WebSymbolsQueryConfigurator`).
The statically contributed symbols are available globally, depending on the context setting of the contributing Web Types file.
If the containing tag is matched to a Web Symbol it is added to the scope for attribute matching.

> For dynamic contributions the `WebSymbolsQueryConfigurator#getScope`'s `element` parameter can be:

Reviewer: ```suggestion
For dynamic contributions the `WebSymbolsQueryConfigurator.getScope()`'s `element` parameter can be:
```

We don't use `#` to refer to methods. If it is clear from the context, we could also leave out the `WebSymbolsQueryConfigurator` and just use `getScope()`.

### CSS

#### Properties {#css-properties}

|---|---|
|||
|Namespace|`css`|
|Kind|`properties`|

Web Symbols, which represent available CSS properties. [Custom CSS properties (variables)](https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_custom_properties) names should be prefixed with `--`.
CSS properties can be contributed statically or dynamically (through `WebSymbolsQueryConfigurator`).
The statically contributed symbols are available globally, depending on the context setting of the contributing Web Types file.

Within a CSS file, the additional scope for matching properties is built from Web Symbols matching HTML element names from terminal selectors from the enclosing ruleset.

Within an HTML element `style` attribute value, the additional scope is built from Web Symbols matching the enclosing HTML element.

Scope for custom properties (variables) references (arguments for [`var()`](https://developer.mozilla.org/en-US/docs/Web/CSS/var) function) is built the same way
as for properties. Only properties with names starting with `--` are taken into account.

> For dynamic contributions the `WebSymbolsQueryConfigurator#getScope`'s `element` parameter can be:

Reviewer: ```suggestion
For dynamic contributions the `WebSymbolsQueryConfigurator.getScope()`'s `element` parameter can be:
```

> - `CssDeclaration` - Web Symbols should represent available CSS properties for matching. *The CSS declaration actual name should not be taken into account when building the scope*.

Reviewer 2: remove italics

Author: I wanted to differentiate this from the rest of the paragraph. If you think it's better without itlics, I guess I need to remove them everywhere.

#### Pseudo-elements {#css-pseudo-elements}

|---|---|
|||
|Namespace|`css`|
|Kind|`pseudo-elements`|

> Web Symbols, which represent available [CSS pseudo-elements](https://developer.mozilla.org/en-US/docs/Web/CSS/Pseudo-elements).

Reviewer 2: Web Symbols representing...

same for all other on this page

Symbols names *should not* be prefixed with `::`.
CSS pseudo-elements can be contributed statically or dynamically (through `WebSymbolsQueryConfigurator`).
The statically contributed symbols are available globally, depending on the context setting of the contributing Web Types file.

The additional scope for matching pseudo-elements is built from Web Symbols matching HTML element name preceding the pseudo-element keyword.

> For dynamic contributions the `WebSymbolsQueryConfigurator#getScope`'s `element` parameter can be:

Reviewer: ```suggestion
For dynamic contributions the `WebSymbolsQueryConfigurator.getScope()`'s `element` parameter can be:
```

- `CssPseudoSelector` - Web Symbols should represent available CSS pseudo-elements for matching. *The CSS pseudo-element actual name should not be taken into account when building the scope*.

Dedicated support for `WebSymbol` interface properties:
- `properties[PROP_ARGUMENTS]` - `true` if pseudo-element keyword accepts arguments.

#### Pseudo-classes {#css-pseudo-classes}

|---|---|
|||
|Namespace|`css`|
|Kind|`pseudo-classes`|

Web Symbols, which represent available [CSS pseudo-classes](https://developer.mozilla.org/en-US/docs/Web/CSS/Pseudo-classes).
Symbols names *should not* be prefixed with `:`.
CSS pseudo-classes can be contributed statically or dynamically (through `WebSymbolsQueryConfigurator`).
The statically contributed symbols are available globally, depending on the context setting of the contributing Web Types file.

The additional scope for matching pseudo-classes is built from Web Symbols matching HTML element name preceding the pseudo-class keyword.

> For dynamic contributions the `WebSymbolsQueryConfigurator#getScope`'s `element` parameter can be:

Reviewer: ```suggestion
For dynamic contributions the `WebSymbolsQueryConfigurator.getScope()`'s `element` parameter can be:
```

- `CssPseudoSelector` - Web Symbols should represent available CSS pseudo-classes for matching. *The CSS pseudo-class actual name should not be taken into account when building the scope*.

Dedicated support for `WebSymbol` interface properties:
- `properties[PROP_ARGUMENTS]` - `true` if pseudo-class keyword accepts arguments.

#### Functions {#css-functions}

|---|---|
|||
|Namespace|`css`|
|Kind|`functions`|

Web Symbols, which represent available [CSS functions](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Functions).
CSS functions can be contributed statically or dynamically (through `WebSymbolsQueryConfigurator`).
The statically contributed symbols are available globally, depending on the context setting of the contributing Web Types file.

> The additional scope for matching functions is built from Web Symbols matching the CSS property name, which value is being calculated.

Reviewer: The last part of this sentence sounds wrong, but I wonder if I understood the meaning correctly. Maybe:

```suggestion
The additional scope for matching functions is built from Web Symbols matching the CSS property name, the value of which is being calculated.
```

> For dynamic contributions the `WebSymbolsQueryConfigurator#getScope`'s `element` parameter can be:

Reviewer: ```suggestion
For dynamic contributions the `WebSymbolsQueryConfigurator.getScope()`'s `element` parameter can be:
```

- `CssFunction` - Web Symbols should represent available CSS functions for matching. *The CSS function actual name should not be taken into account when building the scope*.

#### Classes {#css-classes}

|---|---|
|||
|Namespace|`css`|
|Kind|`classes`|

Web Symbols, which represent available CSS classes.
Symbols names *should not* be prefixed with `.`.
CSS classes can be contributed statically or dynamically (through `WebSymbolsQueryConfigurator`).
The statically contributed symbols are available globally, depending on the context setting of the contributing Web Types file.

Within CSS file the additional scope for matching classes is built from Web Symbols matching HTML element name preceding the class keyword.

Within HTML attribute `class` the additional scope for matching classes is built from Web Symbols matching the enclosing HTML element name.

> For dynamic contributions the `WebSymbolsQueryConfigurator#getScope`'s `element` parameter can be:

Reviewer: ```suggestion
For dynamic contributions the `WebSymbolsQueryConfigurator.getScope()`'s `element` parameter can be:
```

- `CssClass` - Web Symbols should represent available CSS classes for matching. *The CSS class actual name should not be taken into account when building the scope*.
- `XmlAttributeValue` - Web Symbols should represent available CSS classes for matching within the attribute value.

#### Parts {#css-parts}

|---|---|
|||
|Namespace|`css`|
|Kind|`parts`|
|Since|2023.2|

Web Symbols, which represent available HTML element parts for matching with [CSS `::part`](https://developer.mozilla.org/en-US/docs/Web/CSS/::part) pseudo-element.
CSS parts can be contributed statically or dynamically (through `WebSymbolsQueryConfigurator`).
The statically contributed symbols are available globally, depending on the context setting of the contributing Web Types file.

Within CSS file the additional scope for matching classes is built from Web Symbols matching HTML element name preceding the `::part` keyword.

> For dynamic contributions the `WebSymbolsQueryConfigurator#getScope`'s `element` parameter can be:

Reviewer: ```suggestion
For dynamic contributions the `WebSymbolsQueryConfigurator.getScope()`'s `element` parameter can be:
```

- `CssTermImpl` - Web Symbols should represent available CSS parts for matching. *The CSS part actual name should not be taken into account when building the scope*.

### JavaScript

#### String Literals {#js-string-literals}
|---|---|
|||
|Namespace|`js`|
|Kind|`string-literals`|
|Since|2023.2|

Web Symbols, which represent JavaScript or TypeScript string literals available in a particular location.
Only dynamically contributed string literal symbols (through `WebSymbolsQueryConfigurator`) have built-in support.

> The `WebSymbolsQueryConfigurator#getScope`'s `element` parameter can be:

Reviewer: ```suggestion
The `WebSymbolsQueryConfigurator.getScope()`'s `element` parameter can be:
```

- a `JSLiteralExpression`
- an unqualified `JSReferenceExpression`,
  which parent is *not* `JSIndexedPropertyAccessExpression`, `JSCallExpression` or `JSProperty`

#### Properties {#js-properties}
|---|---|
|||
|Namespace|`js`|
|Kind|`properties`|
|Since|2023.2|

Web Symbols, which represent properties of an object, which is a result of a JavaScript or TypeScript expression.
Only dynamically contributed properties symbols (through `WebSymbolsQueryConfigurator`) have built-in support.

> The `WebSymbolsQueryConfigurator#getScope`'s `element` parameter can be:

Reviewer: ```suggestion
The `WebSymbolsQueryConfigurator.getScope()`'s `element` parameter can be:
```

  - a `JSObjectLiteralExpression` - Web Symbols should represent expected properties
  - a `JSExpression` - Web Symbols should represent available properties of the result of the expression. Parent expression
    is `JSReferenceExpression` or `JSIndexedPropertyAccessExpression`

Dedicated support for `WebSymbol` interface properties:
  - `type` - if the type is `JSType` it will be used in JavaScript type evaluator as the type of the property
  - `required` - the JavaScript property is treated as non-optional
  - `properties[PROP_READ_ONLY]` - the JavaScript property is treated as read-only

#### Symbols {#js-symbols}
|---|---|
|||
|Namespace|`js`|
|Kind|`symbols`|
|Since|2023.2|

Web Symbols, which represent JavaScript symbols available for resolve of an unqualified JavaScript reference expression.
Only dynamically contributed symbols (through `WebSymbolsQueryConfigurator`) have built-in support.

> The `WebSymbolsQueryConfigurator#getScope`'s `element` parameter can be:

Reviewer: ```suggestion
The `WebSymbolsQueryConfigurator.getScope()`'s `element` parameter can be:
```

- an unqualified `JSReferenceExpression` - Web Symbols should represent possible symbols to which JavaScript reference could resolve. *The reference actual name should not be taken into account when building the scope.*

Dedicated support for `WebSymbol` interface properties:
- `type` - if the type is `JSType` it will be used in JavaScript type evaluator as the type of the symbol
- `properties[PROP_KIND]` - the kind of the symbol, one of [WebSymbolJsKind] enum values. The kind will be used to render appropriate icon in the code completion popup
:
*Integration with unqualified reference resolution is not available in TypeScript code*

#### DOM Events {#js-events}
|---|---|
|||
|Namespace|`js`|
|Kind|`events`|

Web Symbols, which represent available DOM events.
DOM events can be contributed statically or dynamically (through `WebSymbolsQueryConfigurator`).
The statically contributed symbols are available globally, depending on the context setting of the contributing Web Types file.

The additional scope for matching DOM events is built from Web Symbols matching enclosing HTML element name.

> For dynamic contributions the `WebSymbolsQueryConfigurator#getScope`'s `element` parameter can be:

Reviewer: ```suggestion
For dynamic contributions the `WebSymbolsQueryConfigurator.getScope()`'s `element` parameter can be:
```
